presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java

```diff
@@ -0,0 +1,247 @@
+     * Caller of this private constructor is responsible for making sure `values` is constructed with the same `blockBuilderStatus` as the one in the argument
+     */
+    private ArrayBlockBuilder(BlockBuilderStatus blockBuilderStatus, BlockBuilder values, SliceOutput offsets, SliceOutput valueIsNull)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+        this.values = requireNonNull(values, "values is null");
+        this.offsets = requireNonNull(offsets, "offset is null");
+        this.valueIsNull = requireNonNull(valueIsNull);
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return valueIsNull.size();
+    }
+
+    @Override
+    protected Block getValues()
+    {
+        return values;
+    }
+
+    @Override
+    protected Slice getOffsets()
+    {
+        return offsets.getUnderlyingSlice();
+    }
+
+    @Override
+    protected int getOffsetBase()
+    {
+        return OFFSET_BASE;
+    }
+
+    @Override
+    protected Slice getValueIsNull()
+    {
+        return valueIsNull.getUnderlyingSlice();
+    }
+
+    @Override
+    public void assureLoaded()
+    {
+    }
+
+    private void checkOngoingBlockEntry()
+    {
+        if (ongoingBlockEntry) {
+            throw new IllegalStateException("Cannot write to BlockBuilder between calls to beginBlockEntry and closeEntry");
+        }
+    }
+
+    @Override
+    public BlockBuilder writeByte(int value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeShort(int value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeInt(int value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeLong(long value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeFloat(float value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeDouble(double value)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeBytes(Slice source, int sourceIndex, int length)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public BlockBuilder writeObject(Object value)
+    {
+        checkOngoingBlockEntry();
+
+        Block block = (Block) value;
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                values.appendNull();
+            }
+            else {
+                block.writePositionTo(i, values);
+                values.closeEntry();
+            }
+        }
+
+        if (currentEntrySize != 0) {
+            throw new IllegalStateException("Expected entry size to be exactly " + 0 + " but was " + currentEntrySize);
+        }
+        currentEntrySize++;
```

Comment:
I think you can remove `currentEntrySize` and set `ongoingBlockEntry = true` in this method